lib/matplotlib/axes/_base.py

```diff
@@ -4749,13 +4749,37 @@ def label_outer(self, remove_inner_ticks=False):
         self._label_outer_yaxis(skip_non_rectangular_axes=False,
                                 remove_inner_ticks=remove_inner_ticks)
 
+    def _get_subplotspec_with_optional_colorbar(self):
+        """
+        Return the subplotspec for this Axes, except that if this Axes has been
+        moved to a subgridspec to make room for a colorbar, then return the
+        subplotspec that encloses both this Axes and the colorbar Axes.
+        """
+        ss = self.get_subplotspec()
+        if not ss:
+            return
+        gs = ss.get_gridspec()
+        if (isinstance(gs, mpl.gridspec.GridSpecFromSubplotSpec)
+                and gs.nrows * gs.ncols == 6):
```

Comment:
Please add a comment what this magic is about.

Reply:
I would lean towards the least formal here - the subgridspec trick of  making colorbars is pretty clunky, and note that it is explicitly _not_ used by `layout='constrained'`.  

The original ideas behind grid specs and subgridspecs were very good, but unfortunately loses the relationship between sibling axes if you change add the colorbar and make the colorbar and main axes child subgridspecs.

Reply:
Actually, I found out about the Axes._colorbar attribute, which seems like a much simpler way to retrieve this info.  It was not populated for the tight-layout case, but it seems fine to also register the axes in that case?